core/src/test/scala/org/apache/spark/util/UtilsSuite.scala

```diff
@@ -731,6 +732,52 @@ class UtilsSuite extends SparkFunSuite with ResetSystemProperties {
     assert(!sourceFile2.exists())
   }
 
+  test("SPARK-50716: deleteRecursively - SymbolicLink To File") {
+    val tempDir = Utils.createTempDir()
+    val sourceFile = new File(tempDir, "foo.txt")
+    JFiles.write(sourceFile.toPath, "Some content".getBytes)
+    assert(sourceFile.exists())
+
+    val targetDir = new File(tempDir, "targetDir")
+    assert(targetDir.mkdir())
+
+    val symlinkFile = new File(targetDir, "target.txt")
+    JFiles.createSymbolicLink(symlinkFile.toPath, sourceFile.toPath)
+
+    // Check that the symlink was created successfully
+    assert(JFiles.isSymbolicLink(symlinkFile.toPath))
+    Utils.deleteRecursively(tempDir)
+
+    // Verify that everything is deleted
+    assert(!tempDir.exists)
+    assert(!sourceFile.exists)
+    assert(!targetDir.exists)
+    assert(!symlinkFile.exists)
+  }
+
+  test("SPARK-50716: deleteRecursively - SymbolicLink To Dir") {
```

Comment:
This test case would fail without the current pr  because, without specifying `LinkOption.NOFOLLOW_LINKS`, the `symlinkDir` cannot enter the branch `if (fileAttributes.isDirectory() && !isSymlink(file))`. 

However, `symlinkDir` is not an empty directory too, so there is no logic to delete it, which will result in residuals.